File: rayt-python/rayt_python/hittable.py
import typing

from rayt_python.material import Material
from rayt_python.ray import Ray
from rayt_python.vec3 import Point3, Vec3, dot


class HitRecord:
    def __init__(self, p: Point3, t: float, material: Material) -> None:
        self.p = p
        self.t = t
        self.material = material
        self.front_face = False
        self.normal = Vec3(0.0, 0.0, 0.0)

    def set_face_normal(self, ray: Ray, outward_normal: Vec3) -> None:
        self.front_face = dot(ray.direction, outward_normal) < 0
        self.normal = outward_normal if self.front_face is True else -outward_normal


class Hittable:
    def hit(self, ray: Ray, t_min: float, t_max: float) -> typing.NoReturn:
        raise NotImplementedError

File: rayt-rust/src/hittable.rs
use crate::{
    material::Material,
    ray::Ray,
    vec3::{dot, Point3, Vec3},
};

#[derive(Copy, Clone)]
pub(crate) struct HitRecord {
    pub(crate) p: Point3,
    pub(crate) normal: Vec3,
    pub(crate) material: Box<dyn Material>,
    pub(crate) t: f64,
    pub(crate) front_face: bool,
}

impl HitRecord {
    pub(crate) fn new(p: &Point3, t: f64, material: Box<dyn Material>) -> Self {
        Self {
            p: p.clone(),
            normal: Vec3::default(),
            material,
            t,
            front_face: false,
        }
    }

    pub(crate) fn set_face_normal(mut self, r: &Ray, outward_normal: &Vec3) {
        self.front_face = dot(&r.direction, outward_normal) < 0.0;
        self.normal = if self.front_face {
            *outward_normal
        } else {
            -*outward_normal
        };
    }
}

pub(crate) trait Hittable {
    fn hit(&self, r: &Ray, t_min: f64, t_max: f64) -> Option<HitRecord>;
}
